### PR TITLE
Approach 2: add repo_statistics and gitserver_repos_statistics tables

### PR DESCRIPTION
**Pull request by mrnugget**
This is the 2nd approach to implementing these statistics tables after we discovered that the original approach in https://github.com/sourcegraph/sourcegraph/pull/39660 lead to contention around the `repo_statistics` table.

90% of the code in here is the same as in #39660, what changes is that now we have **multiple rows** in the `repo_statistics` table:

* Every time a `repo` row (and in certain cases: a`gitserver_repo` row) is updated/inserted/deleted, we **append** (!) a row to `repo_statistics` with a diff of the total counts before/after the row change. Example: if a `repo` is deleted we append a row with `total = -1` to the `repo_statistics` table.
* At query time we use `SELECT SUM(total), SUM(cloned), SUM(deleted), ...` to get the current total counts.
* A worker periodically (right now: every 30min) compacts the table by (1) getting the current counts, (2) updating the first row's columns to reflect total counts, (3) deleting all other rows.

## Demo video

https://user-images.githubusercontent.com/1185253/185577375-d6d2d7da-f6a4-4aad-b940-3927a4d7dd6b.mp4

## Test plan

- Existing and new unit tests
- Manual testing

### CODE REVIEW

File: cmd/worker/shared/main.go [hunk 1]
```diff
@@ -47,6 +48,7 @@ func Start(logger log.Logger, additionalJobs map[string]job.Job, registerEnterpr
 		"codeintel-policies-repository-matcher": codeintel.NewPoliciesRepositoryMatcherJob(),
 		"gitserver-metrics":                     gitserver.NewMetricsJob(),
 		"record-encrypter":                      encryption.NewRecordEncrypterJob(),
+		"repo-statistics-compactor":             repostatistics.NewCompactor(),
```
Comment (efritz) [inline cmd/worker/shared/main.go:51]: Can you make sure to add a quick section in `workers.md`?

File: cmd/worker/internal/repostatistics/handler.go [hunk 2]
```diff
@@ -0,0 +1 @@
+package repostatistics
```
Comment (efritz) [inline cmd/worker/internal/repostatistics/handler.go:1]: Is this file purposefully empty?

File: internal/database/repo_statistics.go [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+		rs.Cloned,
+		rs.FailedFetch,
+		rowID,
+	)
+	if err := tx.Exec(ctx, updateQuery); err != nil {
+		return err
+	}
+
+	// Delete all the other rows
+	return tx.Exec(ctx, sqlf.Sprintf(deleteOtherRepoStatisticsForCompactionQueryFmtstr, rowID))
```
Comment (efritz) [inline internal/database/repo_statistics.go:122]: I think this is a bit simpler as a single query:

```sql
WITH
deleted AS (
	DELETE FROM repo_statistics
	RETURNING
		total,
		soft_deleted,
		not_cloned,
		cloning,
		cloned,
		failed_fetch
)
INSERT INTO repo_statistics
SELECT
	SUM(total),
	SUM(soft_deleted),
	SUM(not_cloned),
	SUM(cloning),
	SUM(cloned),
	SUM(failed_fetch)
FROM deleted
```

File: internal/database/repo_statistics.go [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+	}
+	defer func() { err = tx.Done(err) }()
+
+	// Get current state and the first row we have in the database
+	var (
+		rowID int
+		rs    RepoStatistics
+	)
+	row := tx.QueryRow(ctx, sqlf.Sprintf(getRepoStatisticsForCompactionQueryFmtstr))
+	err = row.Scan(&rowID, &rs.Total, &rs.SoftDeleted, &rs.NotCloned, &rs.Cloning, &rs.Cloned, &rs.FailedFetch)
```
Comment (efritz) [inline internal/database/repo_statistics.go:101]: You scan this twice - does this feel like a good opportunity for a basestore scan function (e.g., ScanFirstRepoStatistics)?

**Review by efritz — APPROVED** [review]

File: internal/database/repo_statistics.go (hunk 3 above)
Comment (mrnugget) [reply internal/database/repo_statistics.go:122]: NICE! Thanks!

File: cmd/worker/internal/repostatistics/handler.go (hunk 2 above)
Comment (mrnugget) [reply cmd/worker/internal/repostatistics/handler.go:1]: Uhm, yes, eh... absolutely. 100% intentional. Now _why_ I did that.. well, explanation wouldn't fit in here. So, uhm, instead I'll remove it and save some tedious explaining.

File: cmd/worker/shared/main.go (hunk 1 above)
Comment (mrnugget) [reply cmd/worker/shared/main.go:51]: Yes!

File: internal/database/repo_statistics.go (hunk 4 above)
Comment (mrnugget) [reply internal/database/repo_statistics.go]: Stopped scanning it twice, but still went shopping with the basestore helpers.

**Comment by mrnugget** [comment]
Merged. As always: happy to address any more feedback post-merge.

**Pull request merged**